### PR TITLE
Standalone Lite: Move Lite classes

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #14726.
**This PR should be REBASE MERGED**

Moves the LightningLite class and wrappers to the standalone package. 
There are two commits:

- Moving the files/classes to the lite package.
- Add backward compatibility: PL has a subclass implementation in `src/pytorch_lightning/lite` with:
    - Conversion logic that turns `gpus=x` and `tpu_cores=x` into `devices=x`
    - Conversion logic that turns `plugins=PLPrecisionPlugin` into `plugins=LitePrecisionPlugin`
    - Conversion logic that turns `strategy=PLStrategy` into `strategy=LiteStrategy`

Please review the two commits separately. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @awaelchli @rohitgr7 @carmocca @akihironitta

### CODE REVIEW

File: src/lightning_lite/plugins/precision/__init__.py [hunk 1]
```diff
@@ -11,15 +11,17 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from lightning_lite.plugins.precision.precision import Precision  # isort:skip
```
Comment (awaelchli) [inline src/lightning_lite/plugins/precision/__init__.py:14]: Had to move this to the top, otherwise circular import at import of double

File: src/lightning_lite/connector.py [hunk 2]
```diff
@@ -298,56 +290,12 @@ def _check_device_config_and_set_final_flags(
...
-        # TODO: Delete this method when num_processes, gpus, ipus and tpu_cores gets removed
-        self._map_deprecated_devices_specific_info_to_accelerator_and_device_flag(devices, gpus, tpu_cores)
-
         if self._devices_flag == "auto" and self._accelerator_flag is None:
             raise ValueError(
                 f"You passed `devices={devices}` but haven't specified"
                 " `accelerator=('auto'|'tpu'|'gpu'|'cpu'|'mps')` for the devices mapping."
             )
 
-    def _map_deprecated_devices_specific_info_to_accelerator_and_device_flag(
```
Comment (awaelchli) [inline src/lightning_lite/connector.py:310]: This is now handled in the pl.LightningLite class

File: src/pytorch_lightning/lite/lite.py [hunk 3]
```diff
@@ -65,376 +84,238 @@ class LightningLite(ABC):
...
+                gpus=gpus,
+                tpu_cores=tpu_cores,
+            )
+
+        lite_plugins: Optional[Union[_LITE_PLUGIN_INPUT, List[_LITE_PLUGIN_INPUT]]]
+        if isinstance(plugins, PLPrecisionPlugin):
+            lite_plugins = _to_lite_precision_plugin(plugins)
+        elif isinstance(plugins, list):
+            lite_plugins = [
+                _to_lite_precision_plugin(plugin) if isinstance(plugin, PLPrecisionPlugin) else plugin  # type: ignore[misc]  # noqa: E501
```
Comment (awaelchli) [inline src/pytorch_lightning/lite/lite.py:124]: Mypy has a problem here with the list comprehension that I couldn't resolve. Maybe @carmocca has an idea.

Comment (carmocca) [reply src/pytorch_lightning/lite/lite.py:124]: `_to_lite_precision_plugin` supports returning an Optional, and lite_plugins is meant to be a list of plugins, not a list of optional plugins.

There's multiple ways you can fix it. I would not allow optional returns from `_to_lite_precision_plugin`

File: examples/convert_from_pt_to_pl/image_classifier_2_lite.py [hunk 4]
```diff
@@ -38,10 +38,10 @@
 from torch.optim.lr_scheduler import StepLR
 from torchmetrics.classification import Accuracy
 
-from lightning_lite.lite import LightningLite  # import LightningLite
 from pytorch_lightning import seed_everything
 from pytorch_lightning.demos.boring_classes import Net
 from pytorch_lightning.demos.mnist_datamodule import MNIST
+from pytorch_lightning.lite import LightningLite  # import LightningLite
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply examples/convert_from_pt_to_pl/image_classifier_2_lite.py:44]: So in the first commit, the refactor tool just updated the imports. In the second commit, I then reverted the example changes, as we said there won't be any separate package for 1.8 and so examples should still show the pl import.

File: tests/tests_lite/test_connector.py [hunk 5]
```diff
@@ -262,21 +262,14 @@ def test_accelerator_cpu(*_):
...
     with pytest.raises(
         RuntimeError,
         match="CUDAAccelerator can not run on your system since the accelerator is not available.",
     ):
-        _Connector(accelerator="cuda")
+        _Connector(accelerator="cuda", devices=1)
 
-    with pytest.deprecated_call(match=r"is deprecated in v1.7 and will be removed"):
-        _Connector(accelerator="cpu", gpus=1)
+    # no warnings
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/tests_lite/test_connector.py]: yeah this can be removed

File: src/pytorch_lightning/lite/lite.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+            if gpus:
+                accelerator = "cuda"
+
+    return deprecated_devices_specific_flag, accelerator
+
+
+def _to_lite_strategy(strategy: PLStrategy) -> LiteStrategy:
+    """Re-instantiates a PL-Strategy as the corresponding Lite-Strategy."""
+
+    if strategy.__class__ == PLDDPStrategy:
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/lite/lite.py]: yes, it was just for avoiding squiggly lines
removed

File: src/pytorch_lightning/lite/lite.py (hunk 3 above)
Comment (awaelchli) [reply src/pytorch_lightning/lite/lite.py]: Thanks, yeah that makes sense. addressed

File: src/pytorch_lightning/callbacks/early_stopping.py [hunk 7]
```diff
@@ -261,7 +261,7 @@ def _improvement_message(self, current: Tensor) -> str:
 
     @staticmethod
     def _log_info(trainer: Optional["pl.Trainer"], message: str, log_rank_zero_only: bool) -> None:
-        rank = _get_rank(strategy=(trainer.strategy if trainer is not None else None))
+        rank = _get_rank(strategy=(trainer.strategy if trainer is not None else None))  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/pytorch_lightning/callbacks/early_stopping.py]: done

File: examples/convert_from_pt_to_pl/image_classifier_2_lite.py [hunk 8]
```diff
@@ -38,10 +38,10 @@
 from torch.optim.lr_scheduler import StepLR
 from torchmetrics.classification import Accuracy
 
+from lightning_lite.lite import LightningLite  # import LightningLite
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply examples/convert_from_pt_to_pl/image_classifier_2_lite.py]: For the future, yes, ok. See my reply below

File: src/pytorch_lightning/lite/lite.py [hunk 9]
```diff
@@ -11,39 +11,58 @@
...
+from pytorch_lightning.strategies import DDPSpawnShardedStrategy as PLDDPSpawnShardedStrategy
+from pytorch_lightning.strategies import DDPSpawnStrategy as PLDDPSpawnStrategy
+from pytorch_lightning.strategies import DDPStrategy as PLDDPStrategy
+from pytorch_lightning.strategies import DeepSpeedStrategy as PLDeepSpeedStrategy
+from pytorch_lightning.strategies import SingleDeviceStrategy as PLSingleDeviceStrategy
+from pytorch_lightning.strategies import SingleTPUStrategy as PLSingleTPUStrategy
+from pytorch_lightning.strategies import Strategy as PLStrategy
+from pytorch_lightning.strategies import TPUSpawnStrategy as PLTPUSpawnStrategy
+
+_PluginsTuple = (PLPrecisionPlugin, PLClusterEnvironment, PLCheckpointIO)
```
Comment (carmocca) [inline src/pytorch_lightning/lite/lite.py:59]: Why not?:

```suggestion
_PL_PLUGIN_TUPLE = (PLPrecisionPlugin, PLClusterEnvironment, PLCheckpointIO)
```

File: src/pytorch_lightning/lite/lite.py [hunk 10]
```diff
@@ -65,376 +84,229 @@ class LightningLite(ABC):
...
+            timeout=strategy._timeout,
+            start_method=strategy._start_method,
+            **strategy._ddp_kwargs,
         )
 
-    @staticmethod
-    def _get_distributed_sampler(dataloader: DataLoader, **kwargs: Any) -> DistributedSampler:
-        kwargs.setdefault("seed", int(os.getenv("PL_GLOBAL_SEED", 0)))
-        return DistributedSamplerWrapper(dataloader.sampler, **kwargs)
+    if type(strategy) == PLTPUSpawnStrategy:
```
Comment (carmocca) [inline src/pytorch_lightning/lite/lite.py:211]: classes are singletons and should be checked with is

```suggestion
    if type(strategy) is PLTPUSpawnStrategy:
```

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/lite/lite.py (hunk 9 above)
Comment (awaelchli) [reply src/pytorch_lightning/lite/lite.py]: It's not actually used. I don't remember why I added it. 
Removed

File: src/pytorch_lightning/lite/lite.py (hunk 10 above)
Comment (awaelchli) [reply src/pytorch_lightning/lite/lite.py]: done

File: src/lightning_lite/plugins/precision/__init__.py (hunk 1 above)
Comment (justusschock) [reply src/lightning_lite/plugins/precision/__init__.py:14]: I don't get why it would be circular. What is imported in Double precision that would create the circle? If we just import both of them here it shouldn't create a circle?

**Review by justusschock — APPROVED** [review]

File: src/lightning_lite/plugins/precision/__init__.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning_lite/plugins/precision/__init__.py:14]: If the import is like this in the init file:

(1) from lightning_lite.plugins.precision.other import OtherClassThatSubclassesPrecision
(2) from lightning_lite.plugins.precision.precision import Precision
 
Then while loading the init, it first goes into (1), but other.py tries to import Precision so it goes back to this file and tries to rerun (1), and then you get "cant import blabla due to partially initialized module"

File: src/lightning_lite/connector.py [hunk 11]
```diff
@@ -99,8 +99,6 @@ def __init__(
         num_nodes: int = 1,
         precision: Union[int, str] = 32,
         plugins: Optional[Union[_PLUGIN_INPUT, List[_PLUGIN_INPUT]]] = None,
-        tpu_cores: Optional[Union[List[int], str, int]] = None,  # deprecated
```
Comment (tchaton) [inline src/lightning_lite/connector.py:102]: Neat !

**Review by tchaton — APPROVED** [review]
LGTM !
